docs: migrate 'define metadata' (#5352)
canonical/snapcraft: docs/how-to/craft/configure-distribution-information.rst

.. _how-to-configure-distribution-information:

> Configure distribution information

Author: This is an attempt at a more meaningful term than "metadata". Happy to take alternative suggestions.

Reviewer: "package information" or "package metadata" may work.  Distribution is good, but it may be conflated with Linux distributions.

Author: I like "package", thanks. I'll rename the page after all the other changes are in.

Author: 5564297

information is also known as *metadata* and the snap's *manifest*.

Distribution information is a mixture of required and optional keys. This guide covers
the required information first. For the non-essential keys, review the Top-level
directives in the `snapcraft.yaml reference
<https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-top-level-directives>`_.

Configure the required information
----------------------------------

When you initialize a snap, it populates ``snapcraft.yaml`` with the minimal set of
required keys.

The placeholder values looks similar to this:

.. code-block:: yaml
    :caption: snapcraft.yaml

    name: newtest
    base: core24
    version: '0.1'
    summary: Single-line elevator pitch for your amazing snap
    description: |
      This is my-snap's description. You have a paragraph or two to tell the
      most important story about your snap. Keep it under 100 words though,
      we live in tweetspace and your description wants to look good in the snap
      store.

    grade: devel
    confinement: devmode

    parts:
      my-part:
        plugin: nil

When crafting a snap, fill these keys as follows:

> - For ``name``, give your snap a unique name to distinguish it from all others. The

Author: I feel like this list might be better served as a two-column table. Thoughts?

  and a brief *Getting started* guide are legitimate uses for the description.

- For ``grade``, specify the production readiness of your snap. While developing, leave
  this set to ``devel`` to disable the snapd guardrails. When your snap is ready for
  release, set it to ``stable``.

- For ``confinement``, set how strong the sandboxing of the snap is. A snap's
  confinement level is the degree of isolation it has from the host system. When first
  crafting, leave this as ``devmode`` to disable sandboxing until you have a working
  snap. In the rare case where your snap needs higher levels of system access, like a
  traditional unsandboxed package, you can :ref:`enable classic confinement
  <how-to-enable-classic-confinement>`.

Reuse information
-----------------

When you're crafting a snap for existing software, it's easier to import its
distribution information from an external source instead of manually copying it. There
are two ways to map this data in the ``snapcraft.yaml`` project file. You can:

- Copy the main distribution information from a standard `AppStream`_ metadata file.
- Copy the ``.desktop`` file into an app by reading its component ID.

> - Set the snap's version and grade through a script.

Author: How about: df0d47c

.. code-block:: xml
    :caption: bravo.metainfo.xml

    <?xml version="1.0" encoding="UTF-8"?>
    <component type="desktop-app">
      <id>com.alex.bravo</id>
      <name>Bravo</name>
      <project_license>GPL-3.0+</project_license>
      <summary>Single-line elevator pitch for your amazing app</summary>
      <description>
        This is apps's description. A paragraph or two to tell the most
        important story about it.
      </description>
      <icon type="local">assets/icon.png</icon>
      <launchable type="desktop-id">com.alex.bravo.desktop</launchable>
      <releases>
          <release date="2025-01-01" version="1.5.3"/>
      </releases>
      <update_contact>alex@example.com</update_contact>
      <url type="homepage">https://example.com</url>
      <url type="bugtracker">https://example.com/issues</url>
      <url type="vcs-browser">https://github.com/alex/bravo</url>
      <url type="translate">https://example.com</url>
      <url type="donation">https://example.com/donate</url>
    </component>

If you were packaging this project as a snap, this is the distribution data you'd be

> better served by copying, not replicating manually.

Reviewer 2: Ah good catch, thank you :)

The keys that copy this information are ``adopt-info`` at the start of the project file
and ``parse-info`` in the definition for the main part -- typically the main app.

To copy the info, first remove the distribution keys that you're replacing with the
metadata file. These could be, among others, ``title``, ``description``, ``summary``,
``license``, ``contact``, ``donation``, ``issues``, ``source-code``, ``license``, and
``website``.

Then, set ``adopt-info`` to the name of part that contains the metadata file.

Lastly, in the main part definition, set ``parse-info`` to the path of the metadata

> file. The path is a relative to one of the part's internal directories in the snap

Reviewer 2: > The path is a relative to

Is this supposed to be "relative to" or "a relative to"? If it's the latter, I'm not quite sure what that means (same parent directory?).

Reviewer: Nice catch, I think it should be "relative to".

Author: df0d47c

~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For backward compatibility, component IDs in the AppStream metadata can have a
``.desktop`` suffix. If this is the case for the app you're packaging, you can reuse the
file by sourcing it with a special key in the app's definition in your project file.

First, find the component ID in the metadata file. It should be in a ``launchable`` tag,
contain the same ID as the app itself, and end in ``.desktop``. In our metadata file
example, it was:

.. code-block:: xml
    :caption: sampleapp.metainfo.xml

    <launchable type="desktop-id">com.example.sampleapp.desktop</launchable>

Then, in the app's definition in the project file, set the ``common-id`` key to the
app's component ID, *without* the ``.desktop`` extension.

During build, Snapcraft will now copy the ``.desktop`` file into the app from the part.

See this configuration here:

.. code-block:: yaml
    :caption: snapcraft.yaml
    :emphasize-lines: 7

    name: bravo
    adopt-info: bravo-part

    apps:
      bravo:
        command: bravo
        common-id: com.alex.bravo

    parts:
      bravo-part:
        plugin: dump
        source: http://github.com/alex/bravo.git
        parse-info:

>           - usr/share/metainfo/com.alex.bravo.appdata.xml

Author: You're right, there's overlap. And this second method is really about desktop entries, too. I think I'll move this section into the `.desktop` file section.

@mr-cal Does that sound reasonable to you? In the original docs, these tasks were somewhat jumbled together.

>           - usr/share/metainfo/com.alex.bravo.appdata.xml

Reviewer: Yeah, this is not explained well in the existing documentation and there is a lot of conditionality in the behavior.

There is overlap between the common-id and parsing appstream data.  The desktop ID from a parse-info entry will take priority over the desktop file from the common-id.  So you can specify `common-id`, `parse-info`, or both.  I think splitting it up or mentioning the complexities would help.

There's value in specifying the common-id beyond it determining the desktop file - downstream app stores and desktop environments may use the common-id.

>           - usr/share/metainfo/com.alex.bravo.appdata.xml

Author: Addressed in [dc9f23a](https://github.com/canonical/snapcraft/pull/5352/commits/dc9f23a7cebe09e57558d100bc227f73c016b850).

I added a notice:

> You can simultaneously source information from both the `common-id` key described in this section and the `parse-info` key. However, the desktop entry from `parse-info` takes precedence.

From a script in a part
~~~~~~~~~~~~~~~~~~~~~~~

If you need to procedurally define the snap's ``version`` and ``grade`` keys, you can
set them at build time with the ``craftctl`` command, invoked by a script.

Start by setting ``adopt-info`` to the name of an important part, typically the snap's
main part.

After that, set ``override-pull`` to an inline series of ``craftctl`` commands. The
variables ``version`` and ``grade`` map to the keys with the same names. You can set
them like environment variables with the ``set`` verb. Here you can use any external

> source that's accessible through terminal commands from the host's environment, such as an environment variable
> variable or an API endpoint.

Reviewer 2: ```suggestion
source that's accessible through terminal commands from the host's environment, such as an environment
variable or an API endpoint.
```

Author: df0d47c

> During build, Snapcraft will now set the snap's version and grade based on the values
> from any source you like.

Reviewer 2: ```suggestion
During build, Snapcraft will set the snap's version and grade based on the values
from the source you provided.
```

Author: df0d47c

Here's an example of that configuration:

.. code-block:: yaml
    :caption: snapcraft.yaml
    :emphasize-lines: 2, 9-12

    name: bravo
    adopt-info: bravo-part
    ...

    parts:
      bravo-part:
        plugin: dump
        source: http://github.com/alex/bravo.git
        override-pull: |
          craftctl default
          craftctl set version="1.5.3"
          craftctl set grade="stable"

Configure the desktop entry
---------------------------

> Snaps support the Linux `desktop entry
> <https://specifications.freedesktop.org/desktop-entry-spec/latest>` standard. You can

Reviewer 2: ```suggestion
Snaps support the Linux `desktop entry
<https://specifications.freedesktop.org/desktop-entry-spec/latest>`_ standard. You can
```

Also, does the next sentence ("You can use desktop entry files...") imply that this is supported? If so, we could probably include the link there.

Reviewer 2: I think the original text is just missing the underscore.

Reviewer: Ah, I see.

Author: df0d47c

use desktop entry files to define your snap's entry in the desktop environment's various
app menus and launchers. The file controls the entry's presentation and how it launches.
If configured, snapd will automatically add your snap to the app launcher and menus
during installation.

There are three methods to configure the desktop menu entry:

- Copy the desktop entry file to the ``snap/gui`` directory.
- Point the ``desktop`` key in the app definition to a desktop file in the ``prime``
  directory.
- `Use the desktop entry file
  <how-to-configure-distribution-information-appstream-metadata>`_ from the AppStream
  metadata of your app.

.. important::

    The icon in the desktop entry is separate from the ``icon`` key. The latter is used
    in graphical front ends, like the snap's profile in the Snap Store.

Add a desktop entry file
~~~~~~~~~~~~~~~~~~~~~~~~

> To start, create files named ``<snap-name>.desktop`` and ``<snap-name>.png`` in the
> ``snap/gui/`` directory in your project's source. Replace ``<snap-name>`` with the same

Reviewer 2: ```suggestion
To start, create files named ``<snap-name>.desktop`` and ``<snap-name>.png`` in your project's
``snap/gui/`` directory. Replace ``<snap-name>`` with the same
```

Author: Leaving this one out, because we need to be unambiguous about "source".

value as the ``name`` key in the project file.

For the desktop entry file, enter:

.. code-block:: desktop
    :caption: .desktop file

    [Desktop Entry]
    Exec=<app-name>
    Icon=${SNAP}/meta/gui/<snap-name>.png

Replace ``<app-name>`` with the same name you gave the app in the project file. The
name is case-sensitive.

Assign ``Icon`` to the absolute path of the image file. This path must be the location
of the icon after the snap is installed. Since snapcraft copies all the contents of the

> ``snap/gui/`` folder to ``meta/gui`` in the installation, the absolute path of the icon

Reviewer 2: ```suggestion
``snap/gui/`` folder to ``meta/gui`` during installation, the absolute path of the icon
```

Author: df0d47c

in this arrangement is ``${SNAP}/meta/gui/snapname.png``.

> Read a desktop entry file

Author: I sense that this might be the most common or desireable case. Should I make it the first one?

Author: df0d47c
